Regenerate meshes as binary fines, re-baseline tests

Sbozzolo opened the pull request:
This PR regenerates all the meshes as binary files and changes the defaults of the julia scripts to make sure that they produce the actual meshes used. 

OLD PR DESCRIPTION:

My #449 PR adds 360,888 lines. Of course, this does not really reflect the change in this PR. The reason for so many lines is that it includes msh files.

msh files are text files so git handles them as source code. However, we never really want to look at diff of msh files.

I have locally this change to my gitattributes so that I don't accidentally print a 100k loc git diff, but I think that this might be useful more broadly.

This change tells Git to treat all files with the .msh extension as binary files. This means:

1. No diff display: Git won't show textual diffs for these files
2. No merge attempts: Git won't try to merge conflicts in these files
3. Obscure, but did you know that in some cases git modifies line endings (depending on the operating system and git settings)? Setting msh files are binary means that git won't attempt line ending conversions

In addition to this, this PR regenerates all the meshes to save them as binary files.

cameronrutherford commented:
> However, we never really want to look at diff of msh files.

So long as this assumption holds true, I think these changes are fine.

Is it still easy to inspect the diff if I wanted to?

Sbozzolo commented:
> So long as this assumption holds true, I think these changes are fine.

If the number of total element in a mesh changes, every line of the file will be different, and pretty much any change in the mesh generation leads to a different number of elements.

> Is it still easy to inspect the diff if I wanted to?

Yes, you can `git diff --text` if you need to.

Sbozzolo commented:
Note that an alternative is to directly output binary mesh files. That would reduce storage requirement for meshes (and this won't be needed)

hughcars commented:
I think this is a good idea, but should probably be accompanied by converting the meshes to binary format anyway. This would however require rerunning all the mesh generation files, which will change the meshes a little bit. So in the spheres example, this means changing `mesh.jl` with
```

Sbozzolo commented:
I went through the exercise of updating the examples and found quite a few things.

1. There were no instructions on how to generate the meshes. I assumed the defaults values were used to everything, but for the `cpw` examples we have two variants of each mesh. I had to open the mesh with Gmsh and visually inspect it to reverse engineer the parameter used.
2. Updating the documentation is a chore and doing it properly (ie, regenerating the scaling plots, recomputing certain numbers that are mentioned in the text) requires hours of busywork
3. A couple of filenames were wrong (`surface-M` -> `terminal-M`)
4. The `baseline` script was out of sync with the `runtest.jl` so it did not work out of the box

I tried improving what I could along the way:
- I fixed the sync issue with `baseline` (but we need a more automatic solution)
- I updated the documentation to directly include the files from the `ref` folder instead of manually copy&pasting it.
- I removed references to specific numbers (e.g., 0.7%, 1.2%) and used a more generic "percent-level" 

I could not understand how to reproduce this result:
> However, the bulk dielectric loss postprocessing results, computed from the energies written
to `postpro/cavity_impedance/domain-E.csv`, still give ``Q_d = 1/0.004 = 2.50\times 10^3`` for
every mode as expected.

When I look at `domain-E.csv` I see this
```csv
        m,                 E_elec (J),                  E_mag (J),                  E_cap (J),                  E_ind (J),              E_elec[1] (J),                  p_elec[1],               E_mag[1] (J),                   p_mag[1]
 1.00e+00,        +9.139656208419e-02,        +9.139656939516e-02,        +0.000000000000e+00,        +0.000000000000e+00,        +9.139656208419e-02,        +1.000000000000e+00,        +9.139656939516e-02,        +1.000000000000e+00
 2.00e+00,        +9.139656208419e-02,        +9.139656940175e-02,        +0.000000000000e+00,        +0.000000000000e+00,        +9.139656208419e-02,        +1.000000000000e+00,        +9.139656940175e-02,        +1.000000000000e+00
 ```
 So, I don't know where the 0.004 comes from.

Sbozzolo commented:
I take it back, not all the meshes were created with the default function arguments. I tried reproducing them by tweaking `order` and `refinement`, but I could not get quite the same meshes, so I went with ones that have similar number of degrees of freedom.

hughcars commented:
> I went through the exercise of updating the examples and found a few things.
> 
> 1. There were no instructions on how to generate the meshes. I assumed the defaults values were used to everything, but for the `cpw` examples we have two variants of each mesh. I had to open the mesh with Gmsh and visually inspect it to reverse engineer the parameter used. (And I later on learned that some meshes had not been generated with default values, see post below)
> 2. Updating the documentation is a chore and doing it properly (ie, regenerating the scaling plots, recomputing certain numbers that are mentioned in the text) requires hours of busywork
> 3. A couple of filenames were wrong (`surface-M` -> `terminal-M`)
> 4. The `baseline` script was out of sync with the `runtest.jl` so it did not work out of the box
> 
> I tried improving what I could along the way:
> 
> * I added instructions on how to reproduce the example mesh in each file
> * I fixed the sync issue with `baseline` (but we need a more automatic solution)
> * I updated the documentation to directly include the files from the `ref` folder instead of manually copy&pasting it.
> * I removed references to specific numbers (e.g., 0.7%, 1.2%) and used a more generic "percent-level"
> 
> I could not understand how to reproduce this result:
> 
> > However, the bulk dielectric loss postprocessing results, computed from the energies written
> > to `postpro/cavity_impedance/domain-E.csv`, still give `Q_d = 1/0.004 = 2.50\times 10^3` for
> > every mode as expected.
> 
> When I look at `domain-E.csv` I see this
> 
> ```
>         m,                 E_elec (J),                  E_mag (J),                  E_cap (J),                  E_ind (J),              E_elec[1] (J),                  p_elec[1],               E_mag[1] (J),                   p_mag[1]
>  1.00e+00,        +9.139656208419e-02,        +9.139656939516e-02,        +0.000000000000e+00,        +0.000000000000e+00,        +9.139656208419e-02,        +1.000000000000e+00,        +9.139656939516e-02,        +1.000000000000e+00
>  2.00e+00,        +9.139656208419e-02,        +9.139656940175e-02,        +0.000000000000e+00,        +0.000000000000e+00,        +9.139656208419e-02,        +1.000000000000e+00,        +9.139656940175e-02,        +1.000000000000e+00
> ```
> 
> So, I don't know where the 0.004 comes from.

1. This seems like tech debt, the meshes were not historically being regenerated because of HXT, but moving away from that we ought to have that the args needed to generate them are at least documented in the corresponding `mesh.jl`
2. Agreed, this shouldn't be so complicated. The convergence study for example ought to be automated by loading the config and UniformLevels to generate the runs, which can be dispatched. This would still take time to run, but the data generation and then wrangling afterwards should be automated (already or we should fix)
3. Good example of failures relating to these not be automatically updated too. 
4. Similar to above. 

> I could not understand how to reproduce this result:
> 
> > However, the bulk dielectric loss postprocessing results, computed from the energies written
> > to `postpro/cavity_impedance/domain-E.csv`, still give `Q_d = 1/0.004 = 2.50\times 10^3` for
> > every mode as expected.

This is a bit opaque, but if you read through Pozar the Qd (lossless exterior, lossy dielectric) is a function of omega, electric energy and magnetic energy (eq 6.48 in the textbook). So if those 3 don't change, and looking also at 6.47, P_d is a rescaling of the electric energy in the dielectric, thus doesn't change. So the Qd will be identical if energy and frequencies don't change, hence `2.50\times 10^3`. So the extra loss, which is reducing Q via the formula `Q^-1 = (Qc^-1 + Qd^-1)` makes sense. Then the text that comes afterwards with the specific calculated values from the example that gets reference is putting in Qc and Qd manually calculated and checking the answer.

hughcars commented on test/examples/ref/cpw/wave_uniform/port-S.csv:
```diff
@@ -1,8 +1,8 @@
+ 1.70000000e+01,        -2.626264479970e+01,        -1.143564033665e+02,        -4.736018276713e-02,        +1.599655598561e+02,        -4.895037231784e+01,        -4.904802707569e+00,        -4.790887347833e+01,        -7.765040837252e+00,        -4.735965123638e-02,        +1.599655483972e+02,        -2.608364863795e+01,        -9.785910002922e+01,        -4.784107581767e+01,        -8.11818226 ...
+ 2.00000000e+01,        -1.738398596729e+01,        -1.470809418293e+02,        -1.255553416631e-01,        +1.244674434232e+02,        -4.593184969572e+01,        +4.938896714581e+00,        -4.781949281384e+01,        -4.045350818774e+01,        -1.255544208693e-01,        +1.244674261912e+02,        -1.736585334330e+01,        -1.419383355111e+02,        -4.770312116800e+01,        -4.19622003 ...
+ 2.60000000e+01,        -1.666857053872e+01,        +1.450202306938e+02,        -1.554834376819e-01,        +5.476377339055e+01,        -4.380524309615e+01,        -2.924935652818e+01,        -4.666320843387e+01,        -1.046381960012e+02,        -1.554833404623e-01,        +5.476374795487e+01,        -1.665417611768e+01,        +1.436610295289e+02,        -4.712660945709e+01,        -1.06933719 ...
+ 3.20000000e+01,        -2.784324110881e+01,        -1.257074675959e+02,        -5.986160174635e-02,        -1.637318810533e+01,        -5.104789815483e+01,        -3.286456129192e+01,        -4.581734026175e+01,        -1.783858501957e+02,        -5.986175775304e-02,        -1.637322373880e+01,        -2.735830629742e+01,        -7.911261046332e+01,        -4.618509159414e+01,        -1.77518166 ...
```
The last value here for the phases for instance, looks like a pretty big change. It does appear to be just a change of sign though.

hughcars commented on test/examples/ref/cpw/wave_adaptive/surface-F.csv:
```diff
@@ -1,302 +1,302 @@
+ 3.17000000e+01,        +3.431695046943e-15,        +9.693841562337e-14,        +1.288804851828e-03,        +9.972809335007e-01,        +9.785258489234e-01
+ 3.18000000e+01,        +7.930659411908e-15,        +9.598143996925e-14,        +1.287979709669e-03,        +9.971190860313e-01,        +9.783577585966e-01
+ 3.19000000e+01,        +1.238143320651e-14,        +9.492386765002e-14,        +1.286917049312e-03,        +9.969303809209e-01,        +9.781644748330e-01
+ 3.20000000e+01,        +1.678161810864e-14,        +9.376737265214e-14,        +1.285622322665e-03,        +9.967151946577e-01,        +9.779463706155e-01
```
The first column final value is changed sign, pretty close to zero, but that is a surprise.

hughcars commented on docs/src/examples/rings.md:
```diff
@@ -77,22 +81,19 @@ M_{bb} &= 707.2050\text{ pH}
 [`config["Boundaries"]["Postprocessing"]["Inductance"]`](../config/boundaries.md#boundaries%5B%22Postprocessing%22%5D%5B%22Inductance%22%5D)
 in the configuration file. The resulting postprocessed values are written to
-`postpro/surface-M.csv`:
+`postpro/terminal-M.csv`:
```
The text here is somewhat confusing. What's supposed to be output is `surface-F`, the actual magnetic flux, which can effectively be used to calculate M by dividing by values from `terminal-I.csv`, but isn't symmetric by construction, unlike the `terminal-M.csv` etc which come from energy and are naturally symmetric.

hughcars commented on examples/cylinder/mesh/mesh.jl:
```diff
@@ -1,6 +1,11 @@
+
 using Gmsh: gmsh
 
 """
```
I think the default refinement level should probably be higher, as from memory that's the source of the issues with the periodic datasets.

hughcars reviewed (changes requested):
Read through, the cylinder mesh there's a big issue with the default params, check the frequencies of the output meshes. There's a few other spots I noticed quickly in the changed files too, maybe they can be tightened with tolerances, but I'm very wary about adjusting the cpw tolerances to tighter.
